---
title: "catGenes package"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Manual}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

### The catGenes package is intended to help researchers in phylogenetics and phylogenomics to build a fully concatenated or combined (non-interleaved) dataset by automatically comparing individual DNA alignments.


# Installation
### Developmental using devtools
```{r, eval=FALSE}
devtools::install_github("domingoscardoso/catGenes")
library(catGenes)
```

# Standard format for input DNA alignments
### Before loading the DNA alignments to run catGenes functions, make sure the DNA aligments have the species names format as following:
### (i) Underscores separate the genus and species name from the collector+number+genbank accession
### Genus_species_everythingelse    
### (ii) so... no problem if the genus name is abbreviated
### G_species_everythingelse        

# Usage
```{r, eval=FALSE}
path <- system.file("DNAlignments", package = "catGenes")

ITS <- read.nexus.data(paste0(path, "/Gaya_ITS.nex"))
```


```{r, eval=FALSE}

```

